---
title: "microbiome_trend"
author: "Erica Ryu"
date: "5/17/2023"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 8. Microbiome Trend Test

The purpose of this script is to determine which genera follow the lifestyle trend

## load packages
```{r}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(DescTools)
library(data.table)
library(magrittr)
library(tidyr)
library(purrr)
```

## load data
```{r}
phyloseq_complete <- readRDS("output/ps_complete.rds")

# subset based on extraction kit
qiagen <- subset_samples(phyloseq_complete, Condition == "Qiagen")
```

## set colors
```{r}
fivecolors <- c("darkslateblue", "deepskyblue", "lightblue3", "lightsalmon" , "firebrick")
```

## prep taxa table
```{r}
phy_genus <-  tax_glom(qiagen, "Genus", NArm = FALSE);phy_genus
phy_genus_RA <- transform_sample_counts(phy_genus, function(otu) otu/sum(otu))
otu_table <- as.data.frame(as.matrix(phy_genus_RA@otu_table))

setDT(otu_table, keep.rownames = TRUE)[]
colnames(otu_table)[1] <- "SampleID"

otu_table2 <- otu_table %>% data.frame %>% set_rownames(.$SampleID)
otu_table2 <- otu_table2[,-1]
otu_table2 <- t(otu_table2)
tax_table <- as.data.frame(as.matrix(phy_genus@tax_table))

otu_tax <- merge(otu_table2, tax_table, by = 0, all = TRUE)

# change colnames to genus name
otu_tax$Genus <- replace_na(otu_tax$Genus, "_unclassified")

# and add higher taxonomic level to unclassified
otu_tax$Phylum <- ifelse(is.na(otu_tax$Phylum), otu_tax$Kingdom, otu_tax$Phylum)
otu_tax$Class <- ifelse(is.na(otu_tax$Class), otu_tax$Phylum, otu_tax$Class)
otu_tax$Order <- ifelse(is.na(otu_tax$Order), otu_tax$Class, otu_tax$Order)
otu_tax$Family <- ifelse(is.na(otu_tax$Family), otu_tax$Order, otu_tax$Family)

otu_tax$Genus <- ifelse(otu_tax$Genus == "_unclassified", paste(otu_tax$Family, otu_tax$Genus, sep = ""), otu_tax$Genus)

# rename completely unclassified ASV
otu_tax$Genus <- gsub("NA_unclassified", "Unclassified", otu_tax$Genus)

# make genus rownames
rownames(otu_tax) <- otu_tax$Genus
# and remove taxonomic columns
otu_tax <- subset(otu_tax, select = -c(Row.names, Kingdom, Phylum, Class, Order, Family, Genus))

# transpose
trend_asvs <- as.data.frame(t(otu_tax))

# add lifestyle info
trend_asvs$lifestyle  <- ifelse(grepl("CHE",row.names(trend_asvs)), "Foragers",
                              ifelse(grepl("EUR",row.names(trend_asvs)), "American Industrial", 
                              ifelse(grepl("NEW00",row.names(trend_asvs)) | grepl("NEW10",row.names(trend_asvs)) | grepl("THA",row.names(trend_asvs)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(trend_asvs)) | grepl("NEW11",row.names(trend_asvs)), "Expats", "Recently Settled"))))

# establish levels
trend_asvs$lifestyle <- factor(trend_asvs$lifestyle, ordered = TRUE, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial"))
```

## run trend test for each genus
```{r}
# make empty data frame that will be filled with trend test results
trend_result <- data.frame(matrix(ncol = 3, nrow = length(colnames(trend_asvs))))
colnames(trend_result) <- c("taxa", "p_value", "BHadj_p_value")

trend_result$taxa <- colnames(trend_asvs)

# remove lifestyle row
trend_result <- trend_result[!grepl("lifestyle", trend_result$taxa),]

trend_vector <- c()

for(i in colnames(trend_asvs)){
  obj <- print(JonckheereTerpstraTest(as.formula(paste0(i, "~ lifestyle")),  data=trend_asvs))
  trend_vector = c(trend_vector, obj$p.value)
}
## Note: last iteration of trend test will always error as the error pertains to the lifestyle column. ignore error
```

## run multiple test correction
```{r}
trend_result$p_value <- trend_vector

# run multiple test correction
trend_result$BHadj_p_value <- p.adjust(trend_result$p_value, method = "BH", n = length(trend_result$p_value))
```

## save output of trend test
```{r}
write.csv(trend_result, file = "output/trend_result.csv")
```

## plot microbial abundances against lifestyle
```{r}
# log transform
mk_logit <- function(otu_tax) {
  return(log10(otu_tax+0.0001))}
log_otu_tax <- map_df(otu_tax, .f = mk_logit)

# format
rownames(log_otu_tax) <- rownames(otu_tax)

log_otu_tax <- t(log_otu_tax)

log_otu_tax <- as.data.frame(log_otu_tax)

# subset to significant trend microbes
log_sig_taxa <- log_otu_tax[,c("Streptobacillus","Porphyromonadaceae_unclassified","Granulicatella","Moraxella","Simonsiella","Neisseria","Bacteroidetes_unclassified","Brachymonas","Atopobium")]

# change format
log_sig_taxa$sample_name <- row.names(log_sig_taxa)
log_sig_taxa <- pivot_longer(log_sig_taxa, !sample_name, names_to = "Taxa", values_to = "Rel_abund")

# add lifestyle back
log_sig_taxa$lifestyle <- ifelse(grepl("CHE",log_sig_taxa$sample_name), "Foragers", 
                ifelse(grepl("EUR",log_sig_taxa$sample_name), "American Industrial",
                              ifelse(grepl("NEW00",log_sig_taxa$sample_name) | grepl("NEW10",log_sig_taxa$sample_name) | grepl("THA",log_sig_taxa$sample_name), "Agriculturalists",
                              ifelse(grepl("NEW01",log_sig_taxa$sample_name) | grepl("NEW11",log_sig_taxa$sample_name), "Expats", "Recently Settled"))))

# and set levels for lifestyle and taxa
log_sig_taxa$lifestyle <- factor(log_sig_taxa$lifestyle, ordered = TRUE, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial"))

log_sig_taxa$Taxa<- factor(log_sig_taxa$Taxa, ordered = TRUE, levels=c("Streptobacillus","Porphyromonadaceae_unclassified","Granulicatella","Moraxella","Simonsiella","Neisseria","Bacteroidetes_unclassified","Brachymonas","Atopobium"))

# plot
plot_sig_taxa <- ggplot(log_sig_taxa, aes(x = lifestyle, y = Rel_abund, group = lifestyle), color = black) +
  geom_violin(aes(fill=lifestyle), alpha=0.8) +
  geom_boxplot(width=0.1, color="black", alpha=0.2, outlier.shape = NA) +
  scale_x_discrete(limits=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial")) +
  scale_fill_manual(name=NULL,
                    values=fivecolors,
                    breaks=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats",  "American Industrial")) +
  geom_jitter(size = 1, width=0.15, col="darkgreen") + facet_wrap(~Taxa, ncol = 3, scales = "fixed") +
  labs(y = "Log10 Abundance") +
  theme(strip.text = element_text(face = "italic")) +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.95, vjust=0.95)) +
  theme(legend.position = "none") + xlab("")

ggsave(file="figures/plot_sig_taxa.pdf", plot=plot_sig_taxa, width=7, height=7)
```

## check if lifestyle is associated with sequencing depth
```{r}
read_depth <- data.table(as(sample_data(phy_genus), "data.frame"),
                 TotalReads = sample_sums(phy_genus), keep.rownames = TRUE)

lifestyle_reads <- ggplot(read_depth, aes(x = Lifestyle, y = TotalReads, group = Lifestyle), color = black) +
  geom_violin(aes(fill=Lifestyle), alpha=0.8) +
  geom_boxplot(width=0.1, color="black", alpha=0.2, outlier.shape = NA) +
  scale_x_discrete(limits=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats", "Industrial")) +
  scale_fill_manual(name=NULL,
                    values=fivecolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats", "Industrial"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats",  "American Industrial")) +
  geom_jitter(size = 1, width=0.15, col="darkgreen") + 
  labs(y = "Read Depth") +
  theme(strip.text = element_text(face = "italic")) +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 0.95, vjust=0.95)) +
  theme(legend.position = "none") + xlab("")

kruskal.test(TotalReads ~ Lifestyle, data = read_depth)

ggsave(file="figures/lifestyle_reads.pdf", plot=lifestyle_reads, width=7, height=4)
```

## check if Brachy is correlated with sequencing depth
```{r}
brachy_relabund <- otu_tax[c("Brachymonas"),]

brachy_corr <- as.data.frame(cbind(t(brachy_relabund), read_depth$TotalReads, read_depth$Lifestyle))

colnames(brachy_corr) <- c("Brachy_relabund", "TotalReads", "Lifestyle")

brachy_corr[brachy_corr == 0] <- NA

brachy_corr_plot <- ggplot(brachy_corr, color = black, aes(x = log10(as.numeric(Brachy_relabund)), y = as.numeric(TotalReads))) +
  geom_point(shape = 21, color = "black", size = 3, aes(fill = Lifestyle)) +
  scale_fill_manual(name=NULL,
                    values=fivecolors,
                    breaks=c("Foragers", "RecentlySettled", "Agriculturalists", "Expats", "Industrial"),
                    labels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats", "American Industrial")) +
  geom_smooth(method = "glm") +
  labs(x = "Brachymonas relative abundance",
       y = "Total Read depth")

cor.test(as.numeric(brachy_corr$Brachy_relabund), as.numeric(brachy_corr$TotalReads), alternative = "two.sided", method=c("spearman"))

ggsave(file="figures/brachy_corr_plot.pdf", plot=brachy_corr_plot, width=7, height=4)
```

## check if trends hold without the americans
```{r}
no_am_ps <- subset_samples(qiagen, Lifestyle != "Industrial")

# prep OTU table
phy_genus_no_am <-  tax_glom(no_am_ps, "Genus", NArm = FALSE);phy_genus_no_am
phy_genus_RA_no_am <- transform_sample_counts(phy_genus_no_am, function(otu) otu/sum(otu))
otu_table_no_am <- as.data.frame(as.matrix(phy_genus_RA_no_am@otu_table))

setDT(otu_table_no_am, keep.rownames = TRUE)[]
colnames(otu_table_no_am)[1] <- "SampleID"

otu_table2_no_am <- otu_table_no_am %>% data.frame %>% set_rownames(.$SampleID)
otu_table2_no_am <- otu_table2_no_am[,-1]
otu_table2_no_am <- t(otu_table2_no_am)
tax_table_no_am <- as.data.frame(as.matrix(phy_genus_no_am@tax_table))

otu_tax_no_am <- merge(otu_table2_no_am, tax_table_no_am, by = 0, all = TRUE)

## change colnames to genus name
otu_tax_no_am$Genus <- replace_na(otu_tax_no_am$Genus, "_unclassified")

## and add higher taxonomic level to unclassified
otu_tax_no_am$Phylum <- ifelse(is.na(otu_tax_no_am$Phylum), otu_tax_no_am$Kingdom, otu_tax_no_am$Phylum)
otu_tax_no_am$Class <- ifelse(is.na(otu_tax_no_am$Class), otu_tax_no_am$Phylum, otu_tax_no_am$Class)
otu_tax_no_am$Order <- ifelse(is.na(otu_tax_no_am$Order), otu_tax_no_am$Class, otu_tax_no_am$Order)
otu_tax_no_am$Family <- ifelse(is.na(otu_tax_no_am$Family), otu_tax_no_am$Order, otu_tax_no_am$Family)

otu_tax_no_am$Genus <- ifelse(otu_tax_no_am$Genus == "_unclassified", paste(otu_tax_no_am$Family, otu_tax_no_am$Genus, sep = ""), otu_tax_no_am$Genus)

## rename completely unclassified ASV
otu_tax_no_am$Genus <- gsub("NA_unclassified", "Unclassified", otu_tax_no_am$Genus)

## make genus rownames
rownames(otu_tax_no_am) <- otu_tax_no_am$Genus
## and remove taxonomic columns
otu_tax_no_am <- subset(otu_tax_no_am, select = -c(Row.names, Kingdom, Phylum, Class, Order, Family, Genus))

## transpose
trend_asvs_no_am <- as.data.frame(t(otu_tax_no_am))

## add lifestyle info
trend_asvs_no_am$lifestyle  <- ifelse(grepl("CHE",row.names(trend_asvs_no_am)), "Foragers",
                              ifelse(grepl("NEW00",row.names(trend_asvs_no_am)) | grepl("NEW10",row.names(trend_asvs_no_am)) | grepl("THA",row.names(trend_asvs_no_am)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(trend_asvs_no_am)) | grepl("NEW11",row.names(trend_asvs_no_am)), "Expats", "Recently Settled")))

## establish levels
trend_asvs_no_am$lifestyle <- factor(trend_asvs_no_am$lifestyle, ordered = TRUE, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"))

# run trend test for each genus
## make empty data frame that will be filled with trend test results
trend_result_no_am <- data.frame(matrix(ncol = 3, nrow = length(colnames(trend_asvs_no_am))))
colnames(trend_result_no_am) <- c("taxa", "p_value", "BHadj_p_value")

trend_result_no_am$taxa <- colnames(trend_asvs_no_am)

## remove lifestyle row
trend_result_no_am <- trend_result_no_am[!grepl("lifestyle", trend_result_no_am$taxa),]

trend_vector_no_am <- c()

for(i in colnames(trend_asvs_no_am)){
  obj_no_am <- print(JonckheereTerpstraTest(as.formula(paste0(i, "~ lifestyle")),  data=trend_asvs_no_am))
  trend_vector_no_am = c(trend_vector_no_am, obj_no_am$p.value)
}

# multiple test correction
trend_result_no_am$p_value <- trend_vector_no_am

## run multiple test correction
trend_result_no_am$BHadj_p_value <- p.adjust(trend_result_no_am$p_value, method = "BH", n = length(trend_result_no_am$p_value))
```

## check if trends hold without the foragers
```{r}
forager_ps <- subset_samples(qiagen, Lifestyle != "Forager")

# prep OTU table
phy_genus_forager <-  tax_glom(no_am_ps, "Genus", NArm = FALSE);phy_genus_forager
phy_genus_RA_forager <- transform_sample_counts(phy_genus_forager, function(otu) otu/sum(otu))
otu_table_forager <- as.data.frame(as.matrix(phy_genus_RA_forager@otu_table))

setDT(otu_table_forager, keep.rownames = TRUE)[]
colnames(otu_table_forager)[1] <- "SampleID"

otu_table2_forager <- otu_table_forager %>% data.frame %>% set_rownames(.$SampleID)
otu_table2_forager <- otu_table2_forager[,-1]
otu_table2_forager <- t(otu_table2_forager)
tax_table_forager <- as.data.frame(as.matrix(phy_genus_forager@tax_table))

otu_tax_forager <- merge(otu_table2_forager, tax_table_forager, by = 0, all = TRUE)

## change colnames to genus name
otu_tax_forager$Genus <- replace_na(otu_tax_forager$Genus, "_unclassified")

## and add higher taxonomic level to unclassified
otu_tax_forager$Phylum <- ifelse(is.na(otu_tax_forager$Phylum), otu_tax_forager$Kingdom, otu_tax_forager$Phylum)
otu_tax_forager$Class <- ifelse(is.na(otu_tax_forager$Class), otu_tax_forager$Phylum, otu_tax_forager$Class)
otu_tax_forager$Order <- ifelse(is.na(otu_tax_forager$Order), otu_tax_forager$Class, otu_tax_forager$Order)
otu_tax_forager$Family <- ifelse(is.na(otu_tax_forager$Family), otu_tax_forager$Order, otu_tax_forager$Family)

otu_tax_forager$Genus <- ifelse(otu_tax_forager$Genus == "_unclassified", paste(otu_tax_forager$Family, otu_tax_forager$Genus, sep = ""), otu_tax_forager$Genus)

## rename completely unclassified ASV
otu_tax_forager$Genus <- gsub("NA_unclassified", "Unclassified", otu_tax_forager$Genus)

## make genus rownames
rownames(otu_tax_forager) <- otu_tax_forager$Genus
## and remove taxonomic columns
otu_tax_forager <- subset(otu_tax_forager, select = -c(Row.names, Kingdom, Phylum, Class, Order, Family, Genus))

## transpose
trend_asvs_forager <- as.data.frame(t(otu_tax_forager))

## add lifestyle info
trend_asvs_forager$lifestyle  <- ifelse(grepl("CHE",row.names(trend_asvs_forager)), "Foragers",
                              ifelse(grepl("NEW00",row.names(trend_asvs_forager)) | grepl("NEW10",row.names(trend_asvs_forager)) | grepl("THA",row.names(trend_asvs_forager)), "Agriculturalists",
                              ifelse(grepl("NEW01",row.names(trend_asvs_forager)) | grepl("NEW11",row.names(trend_asvs_forager)), "Expats", "Recently Settled")))

## establish levels
trend_asvs_forager$lifestyle <- factor(trend_asvs_forager$lifestyle, ordered = TRUE, levels=c("Foragers", "Recently Settled", "Agriculturalists", "Expats"))

# run trend test for each genus
## make empty data frame that will be filled with trend test results
trend_result_forager <- data.frame(matrix(ncol = 3, nrow = length(colnames(trend_asvs_forager))))
colnames(trend_result_forager) <- c("taxa", "p_value", "BHadj_p_value")

trend_result_forager$taxa <- colnames(trend_asvs_forager)

## remove lifestyle row
trend_result_forager <- trend_result_forager[!grepl("lifestyle", trend_result_forager$taxa),]

trend_vector_forager <- c()

for(i in colnames(trend_asvs_forager)){
  obj_forager <- print(JonckheereTerpstraTest(as.formula(paste0(i, "~ lifestyle")),  data=trend_asvs_forager))
  trend_vector_forager = c(trend_vector_forager, obj_forager$p.value)
}

# multiple test correction
trend_result_forager$p_value <- trend_vector_forager

## run multiple test correction
trend_result_forager$BHadj_p_value <- p.adjust(trend_result_forager$p_value, method = "BH", n = length(trend_result_forager$p_value))
```